File: src/main/java/cz/linkedlist/TileDownloader.java
package cz.linkedlist;

import java.io.File;

/**
 * @author Martin Macko <https://github.com/LinkedList>
 */
public interface TileDownloader {

    String DEFAULT_FOLDER = "/tmp/sentinel/";
    String DESTINATION_FOLDER_PROP = "${destinationFolder:"+DEFAULT_FOLDER+"}";
    String DOWN_URL = "https://sentinel-s2-l1c.s3.amazonaws.com/";

    void downBand(TileSet tileSet, int band);

    void downBand8A(TileSet tileSet);

    void downProductInfo(TileSet tileSet);

    void downTileInfo(TileSet tileSet);

    void downMetadata(TileSet tileSet);

    default void ensureFolderExists(String destinationFolder) {
        File destFolder = new File(destinationFolder);
        if (destFolder.exists() && destFolder.isDirectory()) {
            return;
        }

        final boolean success = destFolder.mkdirs();
        if(success) {
            return;
        }

        throw new RuntimeException("Cannot create necessary folder for downloading: " + destFolder);
    }

}

File: src/main/kotlin/cz/linkedlist/TileDownloader.kt
package cz.linkedlist

import java.io.File

/**
 * @author Martin Macko <https:></https:>//github.com/LinkedList>
 */
interface TileDownloader {

    fun downBand(tileSet: TileSet, band: Int)

    fun downBand8A(tileSet: TileSet)

    fun downProductInfo(tileSet: TileSet)

    fun downTileInfo(tileSet: TileSet)

    fun downMetadata(tileSet: TileSet)

    fun downContent(tileSet: TileSet, content: TileSet.Contents)

    fun downContent(tileSet: TileSet, contents: Array<TileSet.Contents>)

    fun ensureFolderExists(destinationFolder: String) {
        val destFolder = File(destinationFolder)
        if (destFolder.exists() && destFolder.isDirectory) {
            return
        }

        val success = destFolder.mkdirs()
        if (success) {
            return
        }

        throw RuntimeException("Cannot create necessary folder for downloading: $destFolder")
    }

    fun isDownloaded(what: String): File?

    companion object {

        const val DEFAULT_FOLDER = "/tmp/sentinel/"
        const val DESTINATION_FOLDER_PROP = "\${destinationFolder:$DEFAULT_FOLDER}"
        const val DOWN_URL = "https://sentinel-s2-l1c.s3.amazonaws.com/"
    }

}
